test/common/tls/cert_selector/async_cert_selector.cc

```diff
@@ -0,0 +1,49 @@
+#include "test/common/tls/cert_selector/async_cert_selector.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+Ssl::SelectionResult
+AsyncTlsCertificateSelector::selectTlsContext(const SSL_CLIENT_HELLO&,
+                                              Ssl::CertificateSelectionCallbackPtr cb) {
+  ENVOY_LOG_MISC(info, "debug: select context");
+
+  if (mode_ == "sync") {
+    auto& tls_context = selector_ctx_.getTlsContexts()[0];
+    return {Ssl::SelectionResult::SelectionStatus::Success, &tls_context, false};
+  }
+
+  if (mode_ == "async") {
+    ENVOY_LOG_MISC(info, "debug: select cert async");
+    cb_ = std::move(cb);
+    cb_->dispatcher().post([this] { selectTlsContextAsync(); });
+    return {Ssl::SelectionResult::SelectionStatus::Pending, nullptr, false};
+  }
+
+  if (mode_ == "sleep") {
+    ENVOY_LOG_MISC(info, "debug: select cert sleep");
+    // select cert async after 1000ms
+    cb_ = std::move(cb);
+    selection_timer_ = cb_->dispatcher().createTimer([this] { selectTlsContextAsync(); });
+    selection_timer_->enableTimer(std::chrono::milliseconds(1000));
```

Comment:
To make the tests run faster, please reduce this to something like 10 milliseconds